implants/golem/src/inter/eval.rs

```diff
@@ -0,0 +1,304 @@
+                ast: None,
+            },
+            Ok(res) => EvalResult {
+                messages: Either::Right(res.messages),
+                ast: res.ast,
+            },
+        }
+    }
+
+    pub(crate) fn expression(
+        &self,
+        content: String,
+    ) -> EvalResult<impl Iterator<Item = EvalMessage>> {
+        let file = "expression";
+        Self::err(
+            file,
+            AstModule::parse(file, content, &dialect()).map(|module| self.go(file, module)),
+        )
+    }
+
+
+    pub(crate) fn file_with_contents(
+        &self,
+        filename: &str,
+        content: String,
+    ) -> EvalResult<impl Iterator<Item = EvalMessage>> {
+        Self::err(
+            filename,
+            AstModule::parse(filename, content, &dialect()).map(|module| self.go(filename, module)),
+        )
+    }
+
+    fn run(&self, file: &str, ast: AstModule) -> EvalResult<impl Iterator<Item = EvalMessage>> {
+        let new_module;
+        let module = match self.module.as_ref() {
+            Some(module) => module,
+            None => {
+                new_module = Self::new_module(&self.prelude);
+                &new_module
+            }
+        };
+        let mut eval = Evaluator::new(module);
+        eval.enable_terminal_breakpoint_console();
+        let globals = globals();
+        Self::err(
+            file,
+            eval.eval_module(ast, &globals).map(|v| {
+                if self.print_non_none && !v.is_none() {
+                    println!("{}", v);
+                }
+                EvalResult {
+                    messages: iter::empty(),
+                    ast: None,
+                }
+            }),
+        )
+    }
+
+}
+
+impl LspContext for Context {
+    fn parse_file_with_contents(&self, uri: &LspUrl, content: String) -> LspEvalResult {
+        match uri {
+            LspUrl::File(uri) => {
+                let EvalResult { messages, ast } =
+                    self.file_with_contents(&uri.to_string_lossy(), content);
+                LspEvalResult {
+                    diagnostics: messages.map(Diagnostic::from).collect(),
+                    ast,
+                }
+            }
+            _ => LspEvalResult::default(),
+        }
+    }
+
+    fn resolve_load(&self, path: &str, current_file: &LspUrl) -> anyhow::Result<LspUrl> {
+        let path = PathBuf::from(path);
+        match current_file {
+            LspUrl::File(current_file_path) => {
+                let current_file_dir = current_file_path.parent();
+                let absolute_path = match (current_file_dir, path.is_absolute()) {
+                    (_, true) => Ok(path),
+                    (Some(current_file_dir), false) => Ok(current_file_dir.join(&path)),
+                    (None, false) => Err(ResolveLoadError::MissingCurrentFilePath(path)),
+                }?;
+                Ok(Url::from_file_path(absolute_path).unwrap().try_into()?)
```

Comment:
On mobile so can’t verify but feels weird to wrap an unwrap. Is there a better way?

Reply:
I think this is from their example code.
https://github.com/facebookexperimental/starlark-rust/blob/a31c944bf9f20a67b5911d6110595cb404a36b41/starlark/bin/eval.rs#L295

<img width="1135" alt="image" src="https://user-images.githubusercontent.com/7121375/215306694-40cccaa6-a0b3-479f-a369-a68b32520529.png">
Seems to error when returning in a better way.